### PR TITLE
libsolidity: Remove dead code wrt. VariableDeclaration::canHaveAutoType

### LINKED ISSUE #4667: Remove function VariableDeclaration::canHaveAutoType()
It should not be needed anymore with 0.5.0

### PR DESCRIPTION
**Pull request by christianparpart**
Closes #4667

### CODE REVIEW

File: libsolidity/analysis/ReferencesResolver.cpp [hunk 1]
```diff
@@ -417,9 +417,6 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
 
 		_variable.annotation().type = type;
 	}
-	else if (!_variable.canHaveAutoType())
```
Comment (chriseth) [inline libsolidity/analysis/ReferencesResolver.cpp]: Can you instead put an assert at the very beginning of this function and unindent everything by one step?

File: libsolidity/analysis/ReferencesResolver.cpp [hunk 2]
```diff
@@ -292,134 +292,138 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
 	if (_variable.annotation().type)
 		return;
 
+	if (!_variable.typeName())
+	{
+		// This can still happen in very unusual cases where a developer uses constructs, such as
+		// `var a;`, however, such code will have generated errors already.
```
Comment (chriseth) [inline libsolidity/analysis/ReferencesResolver.cpp:298]: Never state anything without proof ;)
Can you use `m_errorReporter.hasErrors()` here?

Comment (christianparpart) [reply libsolidity/analysis/ReferencesResolver.cpp:298]: well yeah, I tried, but ReferencesResolver seems to be invoked within TypeChecker *before* the actual errors are being generated by constructs such as `var a;` - for example `var_empty_decl_1.sol` is reported properly by the TypeChecker, but the ReferenceResolver comes into play *earlier* (hence, I cannot check for hasErrors() just yet). Do you have any proposals what we could do?

Comment (chriseth) [reply libsolidity/analysis/ReferencesResolver.cpp:298]: I think it's fine - TypeChecker has an assert there.

**Review by chriseth — APPROVED** [review]

**Pull request merged**